luna_soc/gateware/core/timer.py

```diff
@@ -0,0 +1,104 @@
+#
+# This file is part of LUNA.
+#
+# Copyright (c) 2020-2025 Great Scott Gadgets <info@greatscottgadgets.com>
+# SPDX-License-Identifier: BSD-3-Clause
+
+""" A simple timer peripheral. """
+
+from amaranth             import *
+from amaranth.lib         import enum, wiring
+from amaranth.lib.wiring  import In, Out, flipped, connect
+
+from amaranth_soc         import csr, event
+
+
+class Peripheral(wiring.Component):
+    class Enable(csr.Register, access="rw"):
+        """Counter enable"""
+        enable: csr.Field(csr.action.RW, unsigned(1))
+
+    class Mode(csr.Register, access="rw"):
+        """Timer mode. When ``periodic`` is set to 1 the counter will automatically be reset to the reload value."""
+        periodic: csr.Field(csr.action.RW, unsigned(1))
+
+    class Reload(csr.Register, access="rw"):
+        """Reload value of counter."""
+        def __init__(self, width):
+            super().__init__({
+                "value": csr.Field(csr.action.RW, unsigned(width))
+            })
+
+    class Counter(csr.Register, access="r"):
+        def __init__(self, width):
+            """Counter value"""
+            super().__init__({
+                "value": csr.Field(csr.action.R, unsigned(width))
+            })
+
+
+    def __init__(self, *, width):
+        if not isinstance(width, int) or width < 0:
+            raise ValueError("Counter width must be a non-negative integer, not {!r}"
+                             .format(width))
+        if width > 32:
+            raise ValueError("Counter width cannot be greater than 32 (was: {})"
+                             .format(width))
+        self.width   = width
+
+        # registers
+        regs = csr.Builder(addr_width=4, data_width=8)
```

Comment:
Why is `data_width` set to 8 here instead of 32 like in lambdasoc?

Reply:
I think timer is the only csr peripheral that has a > 8 bit register field and I'm trying to keep resource usage down on both gateware and firmware side.

What do you think? Is it worth having a 32-bit data-bus for the csr peripherals if it hardly ever gets used?

Reply:
Fair. Did you check that the resource usage is in fact lower? IIRC for CSR, if a field width is wider than the datapath width, shadow registers are instantiated to provide atomic reads/writes.

Reply:
I did check, there's about a 10% gain on the gateware side and we got back about 2 kilobytes on the firmware!

That said, what should _also_ have done is allow `data_width` specification for the new peripheral constructors in the same way that amaranth-soc's gpio peripheral does it:

https://github.com/amaranth-lang/amaranth-soc/blob/5c43cf58f15d9cd9c69ff83c97997708d386b2dc/amaranth_soc/gpio.py#L260